crates/formats/src/param.rs

```diff
@@ -0,0 +1,561 @@
+pub struct ParamTraits<E: StaticBO = LE, O: OffsetType<E> = Offset64, C: CharType<E> = WChar> {
+    phantom: PhantomData<fn() -> (E, O, C)>,
+}
+impl<E: StaticBO, O: OffsetType<E>, C: CharType<E>> traits::ParamTraits for ParamTraits<E, O, C> {
+    type Endian = E;
+    type Offset = O::T;
+    type OffsetPad = O::Pad;
+    type Char = C::Unit;
+    type Str = C::Str;
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Unaligned, FromZeroes, FromBytes)]
+struct ParamTypeOffset<BO: StaticBO> {
+    unk04: U32<BO>,
+    param_type_offset: U64<BO>,
+    unk_pad: [u8; 20],
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Unaligned, FromZeroes, FromBytes)]
+union ParamTypeBlock<BO: StaticBO> {
+    param_type_buf: [u8; 32],
+    offset: ParamTypeOffset<BO>,
+}
+
+/// Describes
+#[repr(C)]
+#[derive(Clone, Unaligned, FromZeroes, FromBytes)]
+pub struct RowDescriptor<T: traits::ParamTraits> {
+    /// ID of the row. This is unique within the param file.
+    pub id: U32<T::Endian>,
+    pad: T::OffsetPad,
+    /// Offset to the data of the row in the param file.
+    pub data_offset: T::Offset,
+    /// Offset to the name of the row in the param file if present.
+    /// Zero otherwise.
+    pub name_offset: T::Offset,
+}
+
+impl<T: traits::ParamTraits> RowDescriptor<T> {
+    /// Gets the data slice for this row.
+    ///
+    /// If the row size is known for the provided file,
+    /// will return the exact slice corresponding to the row data.
+    /// Otherwise, the returned slice will go on to the end of the file.
+    ///
+    /// Retuns [`None`] if the resulting slice is out-of-bounds.
+    fn data<'a>(&self, file: &'a Param<'a, T>) -> Option<&'a [u8]> {
+        let ofs: usize = self.data_offset.into() as usize;
+        match file.detected_row_size {
+            Some(s) => file.data.get(ofs..ofs + s as usize),
+            None => file.data.get(ofs..),
+        }
+    }
+
+    /// Gets the name of this row if one is present.
+    ///
+    /// Retuns [`None`] if the name is not present or could not be read due to
+    /// out-of-bounds indicies, invalid characters, etc.
+    fn name<'a>(&self, file: &'a Param<'a, T>) -> Option<&'a T::Str> {
+        match self.name_offset.into() as usize {
+            0 => None,
+            ofs => T::Str::read_nt_str(file.data.get(ofs..)?),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Unaligned, FromZeroes, FromBytes)]
+pub struct ParamHeader<BO: StaticBO> {
+    strings_offset: U32<BO>,
+    short_data_offset: U16<BO>,
+    unk006: U16<BO>,
+    paramdef_data_version: U16<BO>,
+    row_count: U16<BO>,
+    param_type_block: ParamTypeBlock<BO>,
+    endianess_flag: u8,
+    format_flags_2d: u8,
+    format_flags_2e: u8,
+    paramdef_format_version: u8,
+}
+impl<BO: StaticBO> ParamHeader<BO> {
+    pub fn header_size(&self) -> usize {
+        let f = self.format_flags_2d;
+        if (f & 3) == 3 || (f & 4) != 0 {
+            0x40
+        } else {
+            0x30
+        }
```

Comment:
Is this a version check? Needs some comments/constants.

Reply:
Clarified code and added comments explaining where the check comes from.